最大流問題や最小カット問題などについて記事を足す

kmyk opened the pull request:
close #127
close #139
close #152 
close #153 

まだ作業中なのですがとりあえずプルリクの形にしておきます。
構成をひっくり返すことになるようなレビューコメントがあるならしてほしいですが、詳しいレビューはまだ待ってください。

プレビュー:

- https://kmyk.github.io/algorithm-encyclopedia-staging/maximum-flow-problem#noredirect
- https://kmyk.github.io/algorithm-encyclopedia-staging/minimum-cut-problem#noredirect
- https://kmyk.github.io/algorithm-encyclopedia-staging/moyasu-umeru-mondai#noredirect
- https://kmyk.github.io/algorithm-encyclopedia-staging/project-selection-problem#noredirect

kmyk committed:
最大流問題の記事の下書きを追加する

TODO コメントがまだいくつか

kmyk committed:
画像ファイルについて準備しておく

kmyk commented on _algorithms/minimum-cut-problem.md:
```diff
@@ -0,0 +1,106 @@
+[^moyasu-umeru-local-name]: 競技プログラミングのコミュニティ外では通用しない名前であることに注意したい。
+[^project-selection-problem-name]: TODO: 出典を探す
+[^cut-set-of-edges]: R. J. ウィルソン. グラフ理論入門. 近代科学社, 2001, [ISBN978-4-76-490296-1](https://iss.ndl.go.jp/api/openurl?isbn=9784764902961).
+[^cut-set-of-vertices]: R. Diestel, [Graph Theory](https://www.springer.com/jp/book/9783662536216), 5th ed. Berlin Heidelberg: Springer-Verlag, 2017. TODO: そうらしいんだけど私はこの本を持ってないので noshi91 に確認する
```
@noshi91 カットの定義に辺集合を使うものと頂点集合を使うものがあるというのを注意に書いておきたいです。Diestel では頂点集合を使っているとのことですが、このように書いてしまって大丈夫そうですか？

<blockquote class="twitter-tweet"><p lang="ja" dir="ltr">Graph Theory (Diestel) の network の cut の定義や Wikipedia の定義 <a href="https://t.co/lUe3JHdrkF">https://t.co/lUe3JHdrkF</a> に従うならば、s-t 最小カットは「頂点を (S, V \ S) に 2 分割して、s ∈ S, t ∉ S としたときに、{uv | u ∈ S, v ∉ S} の重み和を最小化せよ」という問題で、</p>&mdash; 熨斗袋 (@noshi91) <a href="https://twitter.com/noshi91/status/1377091602695430151?ref_src=twsrc%5Etfw">March 31, 2021</a></blockquote>

noshi91 replied:
そのツイートは本気で参照されるために書いていないので、避けて欲しいです

Graph Theory (Diestel) は http://diestel-graph-theory.com/basic.html? ここで無料でダウンロードできるので、おすすめです
無向グラフのカットの定義は 1. The Basics p24、network の cut の定義は 6. Flows p152 です

noshi91 replied:
参考:

Schrijver, A. Combinatorial Optimization Polyhedra and Efficiency. Springer では、

D = (V, A) を有向グラフとする。A の部分集合 B が cut であるとは (V の部分集合 U が存在して、U から出る辺全体 = B となる) ことを指す。s \in U かつ t \notin U のとき、U から出る辺全体のことを s-t cut と呼ぶ。
A の部分集合 B が全ての s-t path と交わるとき、B を s-t diconnecting と呼ぶ。(私の訳)

という感じになっていて別の概念です。

kmyk replied:
ありがとうございます。
Diestel が公開されてるの知りませんでした。確認します。

頂点集合で定義したカット (Schrijver の s-t cut) と辺集合で定義したカット (同 s-t disconnecting) が微妙にずれてる (最小カット問題なら差は無視できるが、最大カット問題だと一方は自明でもう一方はNP完全らしい) のは気付いてましたが、後者に名前が名前が付いてるのは知らなかったです。助かります。なんとなく曖昧にしてしまっていましたが、ふたつが異なることを明記しておくことにします。

kmyk replied:
Diestel を見ながら書き直しました。Schrijver は読めてませんが注釈には挙げておきました。
s-t disconnecting は初めて聞いたと思っていましたが、よく考えたら s-t 非連結化集合という名前で見たことあったのでそのように書きました。

noshi91 replied:
今出ている分でカットの定義が 3 つあります。

1. 辺の部分集合であって、s-t を分断するもの (R.J.ウィルソン グラフ理論入門)
2. 頂点の部分集合であって、s \in S, t \notin S であるもの (Diestel Graph Theory)
3. 辺の部分集合であって、2. のような頂点集合から導かれるもの (Schrijver Combinatorial Optimization)

2 3 はほとんど同じなのでまとめて良いようにも思われ、一方で型が違うので (頂点集合 / 辺集合) 気になります。グラフが非連結だと複数の S に対して同じ辺集合が対応したりするのも面倒です。

kmyk replied:
3 つの定義があってすべて異なることを節を切って書いておきました。

kmyk commented:
燃やす埋めるや PSP に独立なページを与えるのは検討し漏らしていました。
「一方から一方へ帰着可能なだけの異なる問題」だと思えば分けたい気持ちは分かって、しかし距離が近すぎて「具体的な応用テクのひとつ」とか「最小カット問題の問題クラスの部分クラスであって名前が付いたもの」でしかないので埋め込んだ方がよいという気持ちもあります。たとえば「有限体のページに mod 1000000007 の場合の話題がある」のような構造です。

微妙なラインなはずでかなり迷っています。どちらがよいかの意見があれば言ってください。
いま以下のように書かれていて、このような説明の仕方 (帰着可能な異なる問題) でそのまま行くのなら独立なページを与えることになるのかなとは思っています。

> 競技プログラミングのコミュニティにおいて「燃やす埋める問題」という名前で呼ばれている問題は、最小カット問題に帰着できる問題のひとつである。

MiSawa commented:
燃やす埋めるとかPSP自体で記事が立っていたりするのと、コストが非負とは限らない一般の場合でも、劣モジュラになるよう反転出来るならば云々のような話も入れたほうがよさそうだとかを考えると、独立なページを与える価値がありそうな気もします。
ただ、燃やす埋めるとPSPでひたすら同じような話が出てきそうで、どうしたもんかなぁという懸念もあり…

kmyk committed:
「燃やす埋める問題」「project selection problem」を独立したページにする

@MiSawa さんが提案してくれたもの:

> 例えば「強連結成分分解のページに、2-SAT の問題定義と帰着の仕方の解説がある」と同型なことになっている気がします。(が、距離がだいぶ違う気もしなくはないです)

という指摘は正しいように聞こえる。

kmyk commented:
燃やす埋める問題と project selection problem に独立したページを与えました。

ところで、一般の最小カット問題の話題なのに「燃やす埋める問題」という言葉を使っているように見えるブログ記事がいくつかあるように見えて対応に困っています。つまり、「最小カットは「頂点を (S, V \ S) に 2 分割して、s ∈ S, t ∉ S としたときに、{uv | u ∈ S, v ∉ S} の重み和を最小化せよ」という問題」なわけですが、この集合 S のことを「燃やす」と呼んで集合 V \ S のことを「埋める」と呼んで議論しているように見えるブログ記事があります。記号として「燃やす」と「埋める」を使えばなんでも燃やす埋める問題になるわけではないはずなのですが。

noshi91 commented:
燃やす埋めると最小カットが近すぎるのが原因だと思います。
燃やす埋めるの厳密な定義は多分存在しない？はずなので

kmyk committed:
劣モジュラ関数まわりの記事の内容を確認した

kmyk committed:
カットの定義に 2 種類ある話のまわりを更新する

頂点集合として定義される場合と、辺集合 (非連結化集合) として定義される場合とがある。

@noshi91 いつもありがとう

kmyk committed:
lint が落ちたので修正する

kmyk commented on _algorithms/project-selection-problem.md:
```diff
@@ -0,0 +1,62 @@
+
+## 注釈
+
+[^project-selection-problem-name]: TODO: 出典を探す
```
tokoharu さんに問い合わせたところ Algorithm Design (Jon Kleinberg) に乗っているらしいと教えてもらった

kmyk replied:
本はまだ確認できてないんですが、時間がかかりそう (図書館へ行かないといけない) なので、issue として立てておいてこのプルリクでは忘れることにしたい。

kmyk replied:
あとでよむ:

<blockquote class="twitter-tweet"><p lang="ja" dir="ltr">Project Selection Problem (を max flow で解くやつ) の初出，ちらっと調べたらこれな気がしたんだけど，妙に引用が少ないのでよく分からん<a href="https://t.co/meWFbO2zx4">https://t.co/meWFbO2zx4</a></p>&mdash; つる（競プロ） (@theory_and_me) <a href="https://twitter.com/theory_and_me/status/1377819721941151749?ref_src=twsrc%5Etfw">April 2, 2021</a></blockquote>

kmyk committed:
project selection problem の出典についての TODO を削除する

Algorithm Design (Jon Kleinberg) に乗ってるらしいと @tokoharu さんに教えてもらったため。
実は project selection problem もローカルネームである、という可能性は消えたと見てよい。
参考文献に挙げておきたいけど、少なくとも私はまだ本の中身を確認できてないので控えておく。

kmyk committed:
カットの定義に 3 種類あることをきちんと説明する

kmyk committed:
カットを頂点集合で定義するときの記号を A ⊆ V でなく S ⊆ V にする

2 色で塗り分けるので B := V \ A として (A, B) という気持ちで書いていたが、B は書かれないし arc の A となんとなく衝突するので、source の S の方がよい気がする。

kmyk committed:
リンクが壊れていたので修正する

kmyk committed:
最小カット問題関連の記事の細部を調整する

kmyk commented:
@noshi91 @MiSawa ひとまず完成と言える状態になったと思います。レビューよろしくお願いします。

kmyk commented on _algorithms/maximum-flow-problem.md:
```diff
@@ -0,0 +1,45 @@
+  最大流最小カット定理によって最大流問題の解の流量は最小カット問題の解の容量に等しい。
+---
+
+# 最大流問題
```
「最大流問題」と「最大フロー問題」のどちらを使うべきかはかなり迷っています。おすすめがあったら教えてほしい。

kmyk replied:
あとでよむ:

<blockquote class="twitter-tweet"><p lang="ja" dir="ltr">Project Selection Problem (を max flow で解くやつ) の初出，ちらっと調べたらこれな気がしたんだけど，妙に引用が少ないのでよく分からん<a href="https://t.co/meWFbO2zx4">https://t.co/meWFbO2zx4</a></p>&mdash; つる（競プロ） (@theory_and_me) <a href="https://twitter.com/theory_and_me/status/1377819721941151749?ref_src=twsrc%5Etfw">April 2, 2021</a></blockquote>

MiSawa replied:
組合せ最適化、アルゴリズムイントロダクション、グラフ・ネットワーク・組合せ論は「最大フロー問題」でした。

kmyk replied:
ありがとうございます。「最大フロー問題」(ついでに「最大フロー」「最大フロー最小カット定理」) で統一するようにしました。

ちなみに「離散凸解析と最適化アルゴリズム」では「最大流問題」(ただし「最大フロー」「最大フロー最小カット定理」) のようです。

MiSawa commented on _algorithms/minimum-cut-problem.md:
```diff
@@ -0,0 +1,73 @@
+
+## カットの定義
+
+与えられたネットワークの $s$-$t$ カットの定義としては、大きく分けて以下のみっつがある。
```
三つ or 3つのほうがよいと思います。

kmyk replied:
「3 種類」にしました。

「一つ」「1 つ」よりは「ひとつ」の方が読みやすいと思っているのですが、1 でなくて 3 だと微妙なんですよね。
私と同じ好みの人の例: https://medium.com/essays-on-japanese-text-expression/one-7bd6ae2d364f
検索して出てきた NHK の指針 (https://www.nhk.or.jp/bunken/research/kotoba/20191201_4.html) だと「3 つ」になりそうですが、どうせなら「3 種類」の方が読みやすいかなと思います。

MiSawa commented on _algorithms/minimum-cut-problem.md:
```diff
@@ -0,0 +1,73 @@
+最小カット問題を考える際にはどの定義を用いても解の容量は同じであるが、カットの定義としてはすべて異なるものである。
+(1.) の定義と (2.) の定義との違いはカットの容量の最大値について考えれば明らかである。
+(2.) の定義と (3.) の定義とはほとんど同じものであるが、ネットワークが非連結な場合に異なってくる。
+(1.) で定義されるものを (2.) や (3.) で定義されるものから区別したいときには、(1.) で定義されるものを $s$-$t$ 非連結化集合 ($s$-$t$ disconnecting) と呼ぶことがある[^s-t-disconnecting]。
```
s-t disconnecting edge set とかではありませんか? ("非連結化集合" は集合だけれど s-t disconnecting は性質に見えます)

noshi91 replied:
これは私の誤訳だったかもしれません (英語力が無くて判断できない)。
For s, t \in V, it is said to be s-t disconnecting, if B intersects each directed s-t path.

MiSawa replied:
その直前は (s-t でないのは) disconnecting arc set とあり、索引では s-t disconnecting arc set という項目名になっているので、やはり「ほげ集合と呼ぶ」と「ほげであると言う」をノリで同一視しているように思います。

kmyk replied:
s-t disconnecting edge set に修正しました。

MiSawa commented on _algorithms/minimum-cut-problem.md:
```diff
@@ -0,0 +1,73 @@
+
+最小カット問題を考える際にはどの定義を用いても解の容量は同じであるが、カットの定義としてはすべて異なるものである。
+(1.) の定義と (2.) の定義との違いはカットの容量の最大値について考えれば明らかである。
+(2.) の定義と (3.) の定義とはほとんど同じものであるが、ネットワークが非連結な場合に異なってくる。
```
「異なる」が曖昧な気がしていて、もうちょっと具体的に「一般には多対一対応になる」みたいな書き方のほうがよい気がします。あと、有向なので、非連結でなくても多対一になる場合がって、例えば u は接続する辺が (u, s) のみのとき、2. では区別出来る「u が s 側に属するか否か」が 3. では区別できません。

noshi91 replied:
これも自分のミスです、申し訳ない...

noshi91 commented on _algorithms/project-selection-problem.md:
```diff
@@ -0,0 +1,62 @@
+
+## 概要
+
+problem selection problem とは、次の形の問題である:
```
> problem

project

kmyk replied:
うっかりしていました……

noshi91 commented on _algorithms/project-selection-problem.md:
```diff
@@ -0,0 +1,62 @@
+こうしてできるネットワーク上での最小カットの容量を $F$ とすると project selection problem の答えは $\sum_i a_i - F$ となる。
+このネットワークを図示すると以下のようになる。
+
+![project selection problem のネットワーク](assets/img/project-selection-problem.svg)
```
> y_{N-1}, b_{N-1}

y_{M-1}, b_{M-1}

kmyk replied:
直しました。ついでに N と M が異なるかもしれないことが分かりやすいように頂点の位置をずらしておきました。

noshi91 commented on _algorithms/maximum-flow-problem.md:
```diff
@@ -0,0 +1,45 @@
+-   [最小カット問題](/minimum-cut-problem)
+    -   最大流最小カット定理によって最大流問題の解の流量は最小カット問題の解の容量に等しい。
+-   [Dinic 法](/dinic)
+    -   Dinic 法は最大流問題を解くアルゴリズムである。最悪計算量は $O(\lvert V \rvert^2 \cdot \lvert E \rvert)$ だが実用的にはかなり速い。
```
O だとやや違和感がありますが、Θ で評価できるのかが分からない (Dinic の最悪ケースに詳しくないため)。

MiSawa replied:
[これ](https://web.archive.org/web/20101126171902/http://www.topcoder.com/tc?module=Static&d1=tutorials&d2=maxFlowRevisited)の真ん中あたりのネットワークで、k^2=Θ(m), p = Θ(n) となるように k と p を調節すれば良さそうな気がします。

MiSawa replied:
pk=O(m) も必要なのでもうちょっと非自明でした。s 全体と t 全体の前後にノードを一つずつ、計 4 つ追加して、u/v と s/t 間の辺はそいつらを経由させるとかですかね :thinking:

noshi91 reviewed (commented):
カットの定義色々が辺の重みが負の時に事情が変わると思いますが、ネットワークというと非負実数ですか？

MiSawa commented on _algorithms/maximum-flow-problem.md:
```diff
@@ -0,0 +1,45 @@
+## 概要
+
+最大流問題とは、与えられたネットワークのフローであって流量が最大のものを求めるという問題である。
+最大流最小カット定理によって最大流問題の解の流量は[最小カット問題](/minimum-cut-problem)の解の容量に等しい。
```
最大流問題と最小カット問題が双対であることにも言及しておいたほうがよいかもしれません。

kmyk replied:
たしかに書いておきたいです。(あまりよく理解してないから曖昧にして逃げてしまっていました……)

しかし「最大流問題と最小カット問題が双対であること」と「最大流最小カット定理が成り立つこと」の関係を私はあまり理解できてないので教えてほしいです。
「最大流問題と最小カット定理とは双対である。しかも、最大流最小カット定理が成り立つ」なのか「最大流問題と最小カット定理とはある特定の関係 (何？) にある問題である 。最大流最小カット定理が成り立つ。よって最大流問題と最小カット定理とは双対だと言える」なのか、どちらでしょうか？ なんとなく後者なのかなとは思っていますが、自信はないです。

MiSawa replied:
大雑把に言うと、

一般に、線形計画問題のインスタンス(主問題)に対し、その双対問題というものがあり、主も双対も feasible であれば両方 bounded でその最適解が一致することが知られています。(強双対定理)
ところで、任意の容量と s, t 付きのネットワークに対し、それ上の最大流問題の線形計画問題としての定式化(の一つ)と、最小カット問題の線形計画問題としての定式化(の一つ)は、互いに双対になります。
最大流も最小カットも、任意の(s!=tな)インスタンスが feasible なので、強双対定理から両者の最適解は一致します。

です。大抵の場合、強双対定理という大道具は使わず、アルゴリズムを持ってきてその有限性の証明とかで独立に示しますが。
「最大フロー問題と最小カット問題は線形計画問題としての双対問題であり、強双対定理からその最適解が一致する。これは特に最大フロー最小カット定理と呼ばれる。」あたりが普通の説明かなぁと思います。

MiSawa commented on _algorithms/project-selection-problem.md:
```diff
@@ -0,0 +1,62 @@
+## 外部リンク
+
+-   [『燃やす埋める』と『ProjectSelectionProblem』 - とこはるのまとめ](http://tokoharuland.hateblo.jp/entry/2017/11/12/234636)<sup>[archive.org](https://web.archive.org/web/20210401023114/http://tokoharuland.hateblo.jp/entry/2017/11/12/234636)</sup>
+    -   <a class="handle">tokoharu</a> によるブログ記事。燃やす埋める問題という問題ではなく project selection problem という別の問題を利用することを提案している。
```
この記事が PSP なので「という別の」は不自然だと思います。

kmyk replied:
たしかに

MiSawa commented on _algorithms/project-selection-problem.md:
```diff
@@ -0,0 +1,62 @@
+-   $N$ 個のプロジェクト $x_0, x_1, x_2, \dots, x _ {N-1}$ と $M$ 個の機械 $y_0, y_1, y_2, \dots, y _ {M-1}$ がある。
+    プロジェクト $x_i$ を実行すると利益 $a_i$ 円 ($a_i \ge 0$) を産む。
+    機械 $y_j$ は購入に費用 $b_j$ 円 ($b_j \ge 0$) かかる。
+    さらに、プロジェクト $x _ {c_k}$ を実行するためは機械 $y _ {d_k}$ が購入されていなければならないという形の条件が $K$ 個与えらている。
```
これ、条件がコスト有限の形だと PSP ではないんですかね?

kmyk replied:
Wikipedia の記述を信用してよいならば、条件がコスト有限の形だと PSP ではないようです。問題のストーリー的にも条件に有限のコストが付くのは不自然な気がします。
論文があるらしいので後で (かつ merge より前に) 確認しておきます。

kmyk committed:
「最大流」でなく「最大フロー」で統一する

-   `scripts/lint.py` にも足しておいた
-   「最大流量」だけは「最大」+「流量」なのでそのまま

kmyk committed:
"s-t disconnecting" ではなく "s-t disconnecting edge set"

kmyk committed:
project selection problem のネットワークの図を修正する

-   y_{N-1}, b_{N-1} でなくて y_{M-1}, b_{M-1} にする
-   N と M が異なるかもしれないことが分かりやすいように頂点をずらしておく

kmyk committed:
typo: "problem selection problem" ではない

kmyk committed:
「という別の」という句は不自然なので削除

kmyk committed:
s/みっつ/3 種類/

kmyk reviewed (commented):
すぐ直せそうなところは先に直しました。落ち着いて考えないと直せないところはまだです。
